server/etcdserver/api/v3rpc/qmon.go

```diff
@@ -0,0 +1,357 @@
+// Copyright 1016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3rpc
+
+import (
+	"context"
+	"fmt"
+	adt "github.com/shenwei356/countminsketch"
+	pb "go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/server/v3/etcdserver"
+	"go.uber.org/zap"
+	"golang.org/x/time/rate"
+	"io/ioutil"
+	"os"
+	"runtime/debug"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	DefaultTotalMemoryBudget        = 1 * 1024 * 1024 * 1024
+	DefaultThrtottleEnableAtPercent = 10
+	MegaByte                        = 1 * 1024 * 1024
+	DefaultRespSize                 = 4 * 1024 * 1024
+	DefaultResetTimer               = 20 * time.Second
+	DefaultAuditThresholdPercent    = 50
+	SmallReqThreshold               = 8 * 1024
+	LargeReqThreshold               = 64 * 1024 * 1024
+	DefaultEstInterval              = 10 * time.Minute
+)
+
+type QueryType int64
+
+const (
+	QueryTypeUnknown QueryType = iota
+	QueryTypeRange
+)
+
+type Query struct {
+	qid   string
+	qsize uint64
+	qtype QueryType
+}
+
+type QueryMonitor interface {
+	// Start the monitoring
+	Start()
+
+	// UpdateUsage : update counters
+	UpdateUsage(req interface{}, resp interface{}, err error)
+
+	// AdmitReq : Decide if we can admit the request.
+	AdmitReq(req interface{}) bool
+
+	// Stop the monitoring
+	Stop()
+}
+
+type BandwidthMonitor struct {
+	totalMemoryBudget            uint64
+	defaultRespSize              uint64
+	degradedBandwidthBytesPerSec uint64
+	enableAtPercent              uint64
+	enableAtBytes                uint64
+	budgetExhausted              bool
+	resetTimer                   time.Duration
+	estRespSize                  *adt.CountMinSketch
+	qcount                       *adt.CountMinSketch
+	respSizeUpdateTime           time.Time
+	estimateUpdateInterval       time.Duration
+	updateEstimate               bool
+	auditThresholdPercent        uint64
+	auditOn                      bool
+	alwaysOnForLargeReq          bool
+	throttle                     *rate.Limiter
+	mu                           sync.Mutex
```

Comment:
mutex should be put close to the concurrent access fields .

Reply:
access to the controller is being synchronized between the periodic update thread and the admission control thread.